File: 1/2021.07.21/148.Sort List/java/solution.java
class Solution {
    public ListNode sortList(ListNode head) {
        ListNode node = head;
        int length = 0;
        while (node != null) {
            length++;
            node = node.next;
        }
        ListNode dummy = new ListNode();
        dummy.next = head;
        for (int len = 1; len < length; len *= 2) {
            ListNode pre = dummy, cur = dummy.next;
            while (cur != null) {
                ListNode head1 = cur;
                for (int i = 1; i < len && cur.next != null; i++) {
                    cur = cur.next;
                }
                ListNode head2 = cur.next;
                cur.next = null;
                cur = head2;
                for (int i = 1; i < len && cur != null && cur.next != null; i++) {
                    cur = cur.next;
                }
                if (cur != null) {
                    ListNode next = cur.next;
                    cur.next = null;
                    cur = next;
                }
                ListNode merged = merge(head1, head2);
                pre.next = merged;
                while (pre.next != null) {
                    pre = pre.next;
                }
            }
        }
        return dummy.next;
    }

    private ListNode merge(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode();
        ListNode cur = dummy;
        while (l1 != null && l2 != null) {
            if (l1.val <= l2.val) {
                cur.next = l1;
                l1 = l1.next;
            } else {
                cur.next = l2;
                l2 = l2.next;
            }
            cur = cur.next;
        }
        if (l1 != null) {
            cur.next = l1;
        } else {
            cur.next = l2;
        }
        return dummy.next;
    }
}

File: 1/2021.07.21/148.Sort List/python/solution.py
class Solution:
    def sortList(self, head: ListNode) -> ListNode:
        def merge(l1, l2):
            dummy = cur = ListNode()
            while l1 and l2:
                if l1.val <= l2.val:
                    cur.next = l1
                    l1 = l1.next
                else:
                    cur.next = l2
                    l2 = l2.next
                cur = cur.next
            cur.next = l1 or l2
            return dummy.next
        
        node = head
        length = 0
        while node:
            length += 1
            node = node.next
        dummy = ListNode()
        dummy.next = head
        sub_len = 1
        while sub_len < length:
            pre, cur = dummy, dummy.next
            while cur:
                head1 = cur
                for _ in range(1, sub_len):
                    if not cur.next:
                        break
                    cur = cur.next
                head2 = cur.next
                cur.next = None
                cur = head2
                for _ in range(1, sub_len):
                    if not cur or not cur.next:
                        break
                    cur = cur.next
                if cur:
                    cur.next, cur = None, cur.next
                pre.next = merge(head1, head2)
                while pre.next:
                    pre = pre.next
            sub_len *= 2
        return dummy.next
